crates/circuit/src/circuit_instruction.rs

```diff
@@ -110,28 +277,135 @@ impl CircuitInstruction {
         self.clone()
     }
 
+    /// The logical operation that this instruction represents an execution of.
+    #[cfg(not(feature = "cache_pygates"))]
+    #[getter]
+    pub fn operation(&self, py: Python) -> PyResult<PyObject> {
+        operation_type_to_py(py, self)
+    }
+
+    #[cfg(feature = "cache_pygates")]
+    #[getter]
+    pub fn operation(&mut self, py: Python) -> PyResult<PyObject> {
+        Ok(match &self.py_op {
+            Some(op) => op.clone_ref(py),
+            None => {
+                let op = operation_type_to_py(py, self)?;
+                self.py_op = Some(op.clone_ref(py));
+                op
+            }
+        })
+    }
+
     /// Creates a shallow copy with the given fields replaced.
     ///
     /// Returns:
     ///     CircuitInstruction: A new instance with the given fields replaced.
+    #[allow(clippy::too_many_arguments)]
     pub fn replace(
         &self,
         py: Python<'_>,
-        operation: Option<PyObject>,
+        operation: Option<OperationInput>,
         qubits: Option<&Bound<PyAny>>,
         clbits: Option<&Bound<PyAny>>,
+        params: Option<SmallVec<[Param; 3]>>,
+        label: Option<String>,
+        duration: Option<PyObject>,
+        unit: Option<String>,
+        condition: Option<PyObject>,
     ) -> PyResult<Self> {
+        let operation = match operation {
+            Some(operation) => operation,
+            None => match &self.operation {
+                OperationType::Standard(op) => OperationInput::Standard(*op),
+                OperationType::Gate(gate) => OperationInput::Gate(gate.clone()),
+                OperationType::Instruction(inst) => OperationInput::Instruction(inst.clone()),
+                OperationType::Operation(op) => OperationInput::Operation(op.clone()),
```

Comment:
I can do the trait impl for `OperationType` -> `OperationInput` (which would simplifiy this block), but the reverse direction wouldn't be possible because the conversion would throw away information we'd need for constructing a new `CircuitInstruction` in  the case of the pyobject catch all variant, which is when we need to call: `convert_py_to_operation_type`.

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/12459/commits/4a93c83b232d554ea471f74b3c8cb115278ea54a